src/CodeGen_GPU_Host.cpp

```diff
@@ -279,6 +279,25 @@ void CodeGen_GPU_Host<CodeGen_CPU>::visit(const For *loop) {
         // Determine the arguments that must be passed into the halide function
         vector<DeviceArgument> closure_args = c.arguments();
 
+        // Sort the args by the size of the underlying type. This is
+        // helpful for avoiding struct-packing ambiguities in metal,
```

Comment:
I thought it was actually required (not merely helpful) for Metal to work correctly?

Reply:
Metal spec seems pretty clear that struct layout in to natural alignment of the type. Something on our side may have rounded sizes smaller than e.g. 32-bit up before passing them to dev run, but have not had a chance to debug. If that is the case, things would still fail even after sorting by size, but just in getting the right values for the bools. Otherwise it seems to be a bug in Metal. (Assuming the bools are at the end.)